### PR TITLE
Make gossiper generic

### PR DESCRIPTION
**Pull request by aaronbuchwald**
This PR makes the gossiper package generic and decouples it from the `*chain.Transaction` type.

### CODE REVIEW

File: api/jsonrpc/server.go [hunk 1]
```diff
@@ -105,12 +105,9 @@ func (j *JSONRPCServer) SubmitTx(
 	if !rtx.Empty() {
 		return errTransactionExtraBytes
 	}
-	if err := tx.VerifyAuth(ctx); err != nil {
-		return err
-	}
```
Comment (aaronbuchwald) [inline api/jsonrpc/server.go:110]: This can be moved into `PreExecute` within `vm.Submit(...)`

File: api/ws/server.go [hunk 2]
```diff
@@ -248,20 +248,11 @@ func (w *WebSocketServer) MessageCallback() pubsub.Callback {
...
 
-			// Verify tx
-			if w.vm.GetVerifyAuth() {
-				if err := tx.VerifyAuth(ctx); err != nil {
-					w.logger.Error("failed to verify sig",
-						zap.Error(err),
-					)
-					return
-				}
-			}
```
Comment (aaronbuchwald) [inline api/ws/server.go:259]: check `GetVerifyAuth` within `vm.Submit(...)`

File: chain/dependencies.go [hunk 3]
```diff
@@ -26,12 +25,6 @@ type Mempool interface {
 	Size(context.Context) int // bytes
 	Add(context.Context, []*Transaction)
 
-	Top(
-		context.Context,
-		time.Duration,
-		func(context.Context, *Transaction) (cont bool, rest bool, err error),
-	) error
-
```
Comment (aaronbuchwald) [inline chain/dependencies.go:34]: Unused from this package - only used by gossiper

File: chain/transaction_marshaller.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package chain
```
Comment (aaronbuchwald) [inline chain/transaction_marshaller.go:4]: moved from `chain/transaction.go`

File: internal/gossiper/manual.go [hunk 5]
```diff
@@ -10,47 +10,74 @@ import (
...
+	doneGossip           chan struct{}
 }
 
-func NewManual(vm VM) *Manual {
-	return &Manual{
-		vm:         vm,
-		doneGossip: make(chan struct{}),
+func NewManual[T Tx](
+	log logging.Logger,
+	gatherer avametrics.MultiGatherer,
```
Comment (aaronbuchwald) [inline internal/gossiper/manual.go:36]: TODO: remove use of `avametrics` type

File: internal/gossiper/proposer.go [hunk 6]
```diff
@@ -14,21 +14,61 @@ import (
...
+	NodeID() ids.NodeID
+	Proposers(ctx context.Context, diff int, depth int) (set.Set[ids.NodeID], error)
+	IsValidator(context.Context, ids.NodeID) (bool, error)
+}
+
+type Tx interface {
+	ID() ids.ID
+	Expiry() int64
+	Size() int
+}
```
Comment (aaronbuchwald) [inline internal/gossiper/proposer.go:59]: should these be moved back to a shared `dependencies.go` file

**Comment by aaronbuchwald** [comment]
Panic from e2e test:

```
[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> handler/handler.go:458 forwarding sync message to consensus {"nodeID": "NodeID-BpfD3AtRmYbvLbpcsGdWFhCx6mHf7E8af", "messageOp": "chits"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd5d7c2]

goroutine 470 [running]:
github.com/ava-labs/hypersdk/api/jsonrpc.(*JSONRPCServer).LastAccepted(0x1?, 0x1?, 0x0?, 0xc0000429f0)
	/home/runner/work/hypersdk/hypersdk/api/jsonrpc/server.go:121 +0x22
[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> snowman/voter.go:60 finishing poll {"result": "Bag[ids.ID]: (Size = 20)\n    2oTSJL7U2uVBxHbp7mFhDqddCJqozy2PGZAUJ3vFkBBPZei9ik: 10\n    QiRYASfVzp3HgnonQxVz9yQbG3CwGZQWQg1WrPMA8B1KAQJWj: 10"}
[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> snowman/voter.go:77 Snowman engine can't quiesce
reflect.Value.call({0xc0000d8c00?, 0xc00012a660?, 0xc000047550?}, {0x10d3316, 0x4}, {0xc000047698, 0x4, 0xc000047580?})
	/opt/hostedtoolcache/go/1.22.8/x64/src/reflect/value.go:596 +0xca6
reflect.Value.Call({0xc0000d8c00?, 0xc00012a660?, 0x1abe780?}, {0xc000047698?, 0x10?, 0xc00055f610?})
[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> handler/handler.go:483 finished handling sync message {"messageOp": "chits"}
[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> handler/handler.go:458 forwarding sync message to consensus {"nodeID": "NodeID-BpfD3AtRmYbvLbpcsGdWFhCx6mHf7E8af", "messageOp": "pull_query"}
[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> handler/handler.go:483 finished handling sync message {"messageOp": "pull_query"}
[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> handler/handler.go:458 forwarding sync message to consensus {"nodeID": "NodeID-BpfD3AtRmYbvLbpcsGdWFhCx6mHf7E8af", "messageOp": "chits"}
	/opt/hostedtoolcache/go/1.22.8/x64/src/reflect/value.go:380 +0xb9
github.com/gorilla/rpc/v2.(*Server).ServeHTTP(0xc000266cc0, {0x1250370, 0xc000135a10}, 0xc000bd87e0)
	/home/runner/go/pkg/mod/github.com/gorilla/rpc@v1.2.0/v2/server.go:220 +0x711
github.com/ava-labs/avalanchego/vms/rpcchainvm/ghttp.(*Server).HandleSimple(0xc00058ac00, {0x1253638, 0xc000bde660}, 0xc000451b00)
	/home/runner/go/pkg/mod/github.com/ava-labs/avalanchego@v1.11.12-rc.2.0.20241001202925-f03745d187d0/vms/rpcchainvm/ghttp/http_server.go:160 +0x382
github.com/ava-labs/avalanchego/proto/pb/http._HTTP_HandleSimple_Handler({0xf918c0, 0xc00058ac00}, {0x1253638, 0xc000bde660}, 0xc000451a80, 0x0)
	/home/runner/go/pkg/mod/github.com/ava-labs/avalanchego@v1.11.12-rc.2.0.20241001202925-f03745d187d0/proto/pb/http/http_grpc.pb.go:129 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001d5800, {0x1253638, 0xc000bde5d0}, {0x125b5a0, 0xc001100000}, 0xc000bd8480, 0xc000e51e60, 0x19869b8, 0x0)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.62.0/server.go:1383 +0xdd1
google.golang.org/grpc.(*Server).handleStream(0xc0001d5800, {0x125b5a0, 0xc001100000}, 0xc000bd8480)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.62.0/server.go:[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> snowman/voter.go:60 finishing poll {"result": "Bag[ids.ID]: (Size = 20)\n    2oTSJL7U2uVBxHbp7mFhDqddCJqozy2PGZAUJ3vFkBBPZei9ik: 11\n    QiRYASfVzp3HgnonQxVz9yQbG3CwGZQWQg1WrPMA8B1KAQJWj: 9"}
[11-18|14:37:23.805] DEBUG <SLapkLoQd2ZGgwCJAeV9NGJmwcZ3QK7jKxvz9u9eXm4gHzi2m Chain> snowman/voter.go:77 Snowman engine can't quiesce
```

File: vm/vm.go [hunk 7]
```diff
@@ -254,6 +254,9 @@ func (vm *VM) Initialize(
 	ctx, span := vm.tracer.Start(ctx, "VM.Initialize")
 	defer span.End()
 
+	// Set defaults
+	vm.mempool = mempool.New[*chain.Transaction](vm.tracer, vm.config.MempoolSize, vm.config.MempoolSponsorSize)
```
Comment (tsachiherman) [inline vm/vm.go:258]: would it make sense to have the chain package export a series of type aliases to hide the generic instantiation ?

**Review by tsachiherman — APPROVED** [review]
looks ok. few minor comments.

File: chain/transaction_marshaller.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+type TxSerializer struct {
+	ActionRegistry *codec.TypeParser[Action]
+	AuthRegistry   *codec.TypeParser[Auth]
+}
+
+func (s *TxSerializer) Unmarshal(data []byte) ([]*Transaction, error) {
+	p := codec.NewReader(data, consts.NetworkSizeLimit)
+	txCount := p.UnpackInt(true)
+	txs := make([]*Transaction, 0, initialCapacity) // DoS to set size to txCount
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply chain/transaction_marshaller.go:23]: Nice catch!

**Review by tsachiherman — APPROVED** [review]

**Pull request merged**